docs/creating-actions.md

```diff
@@ -0,0 +1,163 @@
+  ~~~
+
+### Create a Rust module
+
+All actions should be defined in their own modules as children of the parent
+[`rrg::action`] module and be named the same as the action itself. This module
+should define idiomatic Rust types corresponding to the argument and result
+messages defined in the `.proto` file. Unfortunately, because `Result` is an
+already established concept in Rust, the type corresponding to the `Result`
+message has to be named differently—in Rust code we use `Item` instead. These
+types should implement `rrg::request::Args` and `rrg::request::Item` traits to
+convert between idiomatic Rust and Protocol Buffers types. Finally, the module
+should implement a `handle` method that executes the action.
+
+For example, when implementing an action called `list_foo`, you should create
+a `crates/rrg/src/action/list_foo.rs` file that looks like this:
+
+  ~~~rust
+  // Copyright 2023 Google LLC
+  //
+  // Use of this source code is governed by an MIT-style license that can be found
+  // in the LICENSE file or at https://opensource.org/licenses/MIT.
+
+  /// Arguments of the `list_foo` action.
+  pub struct Args {
+      // TODO.
+  }
+
+  /// Result of the `list_foo` action.
+  pub struct Item {
+      // TODO.
+  }
+
+  /// Handles invocations of the `list_foo` action.
+  pub fn handle<S>(session: &mut S, args: Args) -> crate::session::Result<()>
+  where
+      S: crate::session::Session,
+  {
+      todo!()
+  }
+
+  impl crate::request::Args for Args {
+
+      type Proto = rrg_proto::list_foo::Args;
+
+      fn from_proto(mut proto: Self::Proto) -> Result<Args, crate::request::ParseArgsError> {
+          todo!()
+      }
+  }
+
+  impl crate::response::Item for Item {
+
+      type Proto = rrg_proto::list_foo::Result;
+
+      fn into_proto(self) -> Self::Proto {
+          todo!()
+      }
+  }
+  ~~~
+
+This file has to be declared as a child of the [`rrg::action`] module and should
+be hidden behind the feature declared earlier:
+
+  ~~~rust
+  #[cfg(feature = "action-list_foo")]
+  pub mod list_foo;
+  ~~~
+
+### Register the action
+
+Finally, the action has to be registered so that requests can actually be routed
+to invoke it.
+
+First, you need to extend the [RRG protocol][3]. Add the new variant to the
+`Action` enum that has the same name as the action. If you are contributing
+upstream, pick and use the first available field number. If you are developing
+an action that is internal to your deployment, use one of the field numbers from
+the reserved range between 1024 and 2048.
+
+Once the Protocol Buffers enum has the new field, add a new variant to the Rust
+`Action` enum defined in the [`rrg::request`] module. The compiler and the tests
+will guide you to towards updating the existing code to cover the new variant in
+all the required branches.
+
+As the last step, update the `dispatch` function in the [`rrg::request`] module
+and route the call to the `handle` function you defined. Remember to guard the
```

Comment:
I think this part could use an example, like you did above. I know it's probably less stable than the parts above, but... xD

Reply:
The problem with the example here is that it too much depends on the surrounding code which is why I skipped it. I think with the link to the module fixed, it is actually pretty obvious what should be done.